runtime/gc_modron_startup/mmparse.cpp

```diff
@@ -559,18 +559,152 @@ xlpSubOptionsParser(J9JavaVM *vm, IDATA xlpIndex, XlpError *xlpError, UDATA *req
+		qualifiedSize(&oldSize, &oldQualifier);
+		qualifiedSize(&newSize, &newQualifier);
+		if (0 == (J9PORT_VMEM_PAGE_FLAG_NOT_USED & requestedPageFlags)) {
+			oldPageType = getPageTypeString(requestedPageFlags);
+		}
+		if (0 == (J9PORT_VMEM_PAGE_FLAG_NOT_USED & pageFlags)) {
+			newPageType = getPageTypeString(pageFlags);
+		}
+		if (NULL == oldPageType) {
+			if (NULL == newPageType) {
+				j9nls_printf(PORTLIB, J9NLS_INFO, J9NLS_GC_OPTIONS_LARGE_PAGE_SIZE_NOT_SUPPORTED, oldSize, oldQualifier, newSize, newQualifier);
+			} else {
+				j9nls_printf(PORTLIB, J9NLS_INFO, J9NLS_GC_OPTIONS_LARGE_PAGE_SIZE_NOT_SUPPORTED_WITH_NEW_PAGETYPE, oldSize, oldQualifier, newSize, newQualifier, newPageType);
+			}
+		} else {
+			if (NULL == newPageType) {
+				j9nls_printf(PORTLIB, J9NLS_INFO, J9NLS_GC_OPTIONS_LARGE_PAGE_SIZE_NOT_SUPPORTED_WITH_REQUESTED_PAGETYPE, oldSize, oldQualifier, oldPageType, newSize, newQualifier);
+			} else {
+				j9nls_printf(PORTLIB, J9NLS_INFO, J9NLS_GC_OPTIONS_LARGE_PAGE_SIZE_NOT_SUPPORTED_WITH_PAGETYPE, oldSize, oldQualifier, oldPageType, newSize, newQualifier, newPageType);
+			}
+		}
+	}
+}
+
+/**
+ * 
+ * Obtains and sets the largest page size available on the system. 
+ * Returns true if the operation was successful.
+ * 
+ */
+bool
+gcSetPreferredPageSize(J9JavaVM *vm, MM_GCExtensions *extensions)
+{
+	UDATA defaultLargePageSize = 0;
+	UDATA defaultLargePageFlags = J9PORT_VMEM_PAGE_FLAG_NOT_USED;
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	/* Get the default large page size of non pageable */
+	j9vmem_default_large_page_size_ex(0, &defaultLargePageSize, &defaultLargePageFlags);
+	
+	if (0 != defaultLargePageSize) {
+		extensions->requestedPageSize = defaultLargePageSize;
+		extensions->requestedPageFlags = defaultLargePageFlags;
+	} else {
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * 
+ *  Sets the large page size used within the objectheap when the options -XX:+UseLargePages or -XX:LargePageSizeInBytes=<size> are used.
+ *  The value of the large page size will only be overwritten when these arguments are located in the right most index.
+ *  This is compared to the indices of -Xlp, -Xlp<size>, and -Xlp:objectheap:pagesize=<xx>/
+ * 
+ */
+static bool
+gcProcessXXLargePageObjectHeapArguments(J9JavaVM *vm)
+{
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	MM_GCExtensions *extensions = MM_GCExtensions::getExtensions(vm);
+	if (-1 == vm->largePageSizeRequested) {
+
+		UDATA defaultLargePageSize = 0;
+		UDATA defaultLargePageFlags = J9PORT_VMEM_PAGE_FLAG_NOT_USED;
+
+		/* Get the default large page size of non pageable */
+		j9vmem_default_large_page_size_ex(0, &defaultLargePageSize, &defaultLargePageFlags);
+		
+#if	defined(J9ZOS390)
+		/* Get the default large page size of pageable (ZOS Only) */
+		UDATA defaultLargePageSizePageable = 0;
+		UDATA defaultLargePageFlagsPageable = J9PORT_VMEM_PAGE_FLAG_PAGEABLE;
+
+		j9vmem_default_large_page_size_ex(0, &defaultLargePageSizePageable, &defaultLargePageFlagsPageable);
```

Comment:
This is not good. You are trying to solve problem on wrong level. And even this ugly attempt would not help you because  `j9vmem_default_large_page_size_ex()` on ZOS64 returns 1m pageable if   `OMRPORT_VMEM_MEMORY_MODE_EXECUTE` flag is set only. So Port Library layer should be cleared first.
Also please note that this helper returns page size 0 on ZOS31 (and original helper returns message "option is not supported").
Looks like even existing code has a problem: if `-Xlp` is provided without specifying page size this helper returns 4k fixed (expected to be 1m fixed). I found it on paper and real behaviour should be double checked. This is not a recent regression.
I believe the right way is cleaning Port Library code:
- investigate and document desired changes for `-Xlp` of all forms behaviour, spot methods should be changed
- approve this changes with ZOS experts
- replace all rudiments of using 1m fixed pages as default large page (done for compatibility with old old `-Xlp`)
- test it on machine with variety of LP available

Reply:
>  Looks like even existing code has a problem: if -Xlp is provided without specifying page size this helper returns 4k fixed (expected to be 1m fixed). I found it on paper and real behaviour should be double checked. This is not a recent regression.

I can confirm this is not the case. Using a ZOS machine configured for large pages shows that 1M nonpageable large pages are used.

Reply:
> Also please note that this helper returns page size 0 on ZOS31 (and original helper returns message "option is not supported").

The official docs state that Xlp is not supported on Z31. I don't intend to modify existing behaviour. My expectation is that these two options won't be supported on Z31, we can address it in the future if that's desired.